File: Binary_search/up_bound.cpp
#include<bits/stdc++.h>
using namespace std;

int Mindolit_up_bound_binary_search(int *array,int target, int start ,int end){
  //주어진 값보다 큰값이 나오는 인덱스를 리턴한다.
  while(start<end){
    int mid=(start+end)/2;
    if(array[mid]>target){
      end=mid;
    }
    else if(array[mid]<=target){
      start=mid+1;
    }
  }
  return end;
}

int main(){
  int array[11]={1,1,1,2,2,2,3,3,3,4,6};
  int target;
  cin>>target;
  cout<<Mindolit_up_bound_binary_search(array,target,0,sizeof(array)/sizeof(int));
  system("pause");
}

File: Binary_search/up_bound.py
def up_bound(arr,target,start,end,cur=0): 
  if start>=end:
    return cur
  else:
    mid=int((start+end)/2)
    if arr[mid]<=target:
      cur=mid
      return up_bound(arr,target,mid+1,end,cur)
    elif arr[mid]>target:
      return up_bound(arr,target,start,mid,cur)

if __name__=='__main__':
  sorted_array=[10, 10, 10, 13, 20, 20 ,20]
  print(up_bound(sorted_array,10,0,len(sorted_array)))
